internal/mempool/v1/mempool.go

```diff
@@ -450,422 +416,356 @@ func (txmp *TxMempool) Update(
-	for {
-		if bytes.Equal(tx, wtx.tx) {
-			// We've found a tx in the recheck list that matches the tx that we
-			// received from the ABCI application.
-			// Break, and use this transaction for further checks.
-			break
-		}
-
-		txmp.logger.Error(
-			"re-CheckTx transaction mismatch",
-			"got", wtx.tx.Hash(),
-			"expected", types.Tx(tx).Key(),
-		)
 
-		if txmp.recheckCursor == txmp.recheckEnd {
-			// we reached the end of the recheckTx list without finding a tx
-			// matching the one we received from the ABCI application.
-			// Return without processing any tx.
-			txmp.recheckCursor = nil
+	numLeft := atomic.AddInt64(&txmp.txRecheck, -1)
+	if numLeft <= 0 {
+		defer txmp.notifyTxsAvailable()
+		if numLeft < 0 {
 			return
 		}
-
-		txmp.recheckCursor = txmp.recheckCursor.Next()
-		wtx = txmp.recheckCursor.Value.(*WrappedTx)
 	}
 
-	// Only evaluate transactions that have not been removed. This can happen
-	// if an existing transaction is evicted during CheckTx and while this
-	// callback is being executed for the same evicted transaction.
-	if !txmp.txStore.IsTxRemoved(wtx.hash) {
-		var err error
-		if txmp.postCheck != nil {
-			err = txmp.postCheck(tx, checkTxRes.CheckTx)
-		}
-
-		if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
-			wtx.priority = checkTxRes.CheckTx.Priority
-		} else {
-			txmp.logger.Debug(
-				"existing transaction no longer valid; failed re-CheckTx callback",
-				"priority", wtx.priority,
-				"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
-				"err", err,
-				"code", checkTxRes.CheckTx.Code,
-			)
-
-			if wtx.gossipEl != txmp.recheckCursor {
-				panic("corrupted reCheckTx cursor")
-			}
+	txmp.metrics.RecheckTimes.Add(1)
+	tx := types.Tx(req.GetCheckTx().Tx)
 
-			txmp.removeTx(wtx, !txmp.config.KeepInvalidTxsInCache)
-		}
+	// Find the transaction reported by the ABCI callback. It is possible the
+	// transaction was evicted during the recheck, in which case the transaction
+	// will be gone.
+	elt, ok := txmp.txByKey[tx.Key()]
+	if !ok {
+		return
 	}
+	wtx := elt.Value.(*WrappedTx)
 
-	// move reCheckTx cursor to next element
-	if txmp.recheckCursor == txmp.recheckEnd {
-		txmp.recheckCursor = nil
-	} else {
-		txmp.recheckCursor = txmp.recheckCursor.Next()
+	// If a postcheck hook is defined, call it before checking the result.
+	var err error
+	if txmp.postCheck != nil {
+		err = txmp.postCheck(tx, checkTxRes.CheckTx)
 	}
 
-	if txmp.recheckCursor == nil {
-		txmp.logger.Debug("finished rechecking transactions")
-
-		if txmp.Size() > 0 {
-			txmp.notifyTxsAvailable()
-		}
+	if checkTxRes.CheckTx.Code == abci.CodeTypeOK && err == nil {
+		wtx.SetPriority(checkTxRes.CheckTx.Priority)
+		return // N.B. Size of mempool did not change
 	}
 
+	txmp.logger.Debug(
+		"existing transaction no longer valid; failed re-CheckTx callback",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"err", err,
+		"code", checkTxRes.CheckTx.Code,
+	)
+	txmp.removeTxByElement(elt)
```

Comment:
Do we consider this transaction evicted? Should we record this in our metrics?

Reply:
We can have the metric semantics work one of two ways:
1. Rejected is any event that prevents a transaction from entering the mempool either because of the app or because of Tendermint and Evicted is any action after a transaction enters the mempool and causes it to be forced out.
2. Rejected is any tx that is invalid (from the application's perspective), Evicted is for those that are removed after entering and we probably need another metric to capture transactions that never made it for non-validity based reasons.

For (1) we should call this Evicted for (2) we should call is Rejected. Either is fine so long as it's documented and clear :)

Reply:
Unfortunately, the doc comments in the metrics type don't really allow either: "Rejected" is "transactions that passed CheckTx but failed to make it into the mempool due to resource limits"; "Evicted" is "valid transactions that passed CheckTx and existed in the mempool but were later evicted to make room for higher priority valid transactions that passed CheckTx".

We could go either way, but we have to violate the definitions in any case.  I think extending the definition of Rejected is probably the lest-invasive change here.

Reply:
Er, sorry, I said that backward, I mean "Evicted" is probably the least invasive.

Reply:
So there are no metrics for transactions that failed CheckTx?

Reply:
There appears to be a FailedTxs metric. The original code wasn't setting it, but I will add that here since it makes sense to use what we already have.

The only case I think we don't distinguish right now is "failed on recheck".

Reply:
(I have swapped it to that)

Reply:
Yeah that makes sense to me as well

Reply:
So we have rejected, failed and evicted.

Reply:
Correct. As of now:

- Rejected: Passed initial CheckTx, rejected due to capacity.
- Evicted: Passed initial CheckTx and added to mempool, later ejected either by pruning (for TTL) or because a higher-priority transaction took precedence.
- Failed: Failed either initial or recheck CheckTx.